File: numericalFunctions/nf_specialFunctions/Python/Test/UnitTesting/exponentialIntegral/exponentialIntegralTest.py
#
# <<BEGIN-copyright>>
# Copyright 2021, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>
#

import os
import sys
import glob

import numericalFunctions

# Deployment via Makefiles places shared libraries in .../fudge/numericalFunctions/lib
if len(glob.glob(os.path.join(numericalFunctions.__path__[0], 'lib', '*specialFunctions*'))) > 0:
    from numericalFunctions.lib import specialFunctions

# Deployment via `pip install` places shared libraries in .../site-packages/numericalFunctions
else:
    from numericalFunctions import specialFunctions

options = []
if( 'CHECKOPTIONS' in os.environ ) : options = os.environ['CHECKOPTIONS'].split( )
for argv in sys.argv[1:] : options += [ argv ]

if( '-e' in options ) : print( __file__ )

f = open( '../../../../Test/UnitTesting/exponentialIntegral/exponentialIntegralTest.dat' )
ls = f.readlines( )
f.close( )

errors = 0
for l in ls :
    if( ';' in l ) : break
    if( ',' not in l ) : continue
    n, x, f = l.split( '{' )[1].split( '}' )[0].split( ',' )
    n = int( n )
    x = float( x )
    f = float( f )
    En = specialFunctions.exponentialIntegral( n, x )
    r = 1
    if( f != 0 ) : r = En / f - 1
    if( abs( r ) > 3e-14 ) :
        print( n, x, f, En, r )
        errors += 1

if( errors > 0 ) : raise Exception( "FAILED: %s" % __file__ )

File: numericalFunctions/nf_specialFunctions/Test/UnitTesting/exponentialIntegral/exponentialIntegralTest.c
/*
# <<BEGIN-copyright>>   
# Copyright 2021, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>> 
*/

#include <stdio.h>
#include <stdlib.h>

#include <nf_utilities.h>
#include <nfut_utilities.h>
#include <nf_specialFunctions.h>

struct nXEnOfX {
    int n;
    double x, En;
};

#include "exponentialIntegralTest.dat"

#define nPowerErrors 10
static int verbose = 0;

/*
************************************************************
*/
int main( int argc, char **argv ) {

    int i1, i2, nData = sizeof( data ) / sizeof( data[0] ), counts = 0, iarg, echo = 0, powerErrors[nPowerErrors+1], info = 0;
    double En, r, r2;
    nfu_status status;
    statusMessageReporting smr;

    smr_initialize( &smr, smr_status_Ok );

    for( iarg = 1; iarg < argc; iarg++ ) {
        if( strcmp( "-v", argv[iarg] ) == 0 ) {
            verbose = 1; }
        else if( strcmp( "-e", argv[iarg] ) == 0 ) {
            echo = 1; }
        else if( strcmp( "-i", argv[iarg] ) == 0 ) {
            info = 1; }
        else {
            nfu_printErrorMsg( "ERROR %s: invalid input option '%s'", __FILE__, argv[iarg] );
        }
    }
    if( echo ) printf( "nf_specialFunctions: %s\n", __FILE__ );

    for( i2 = 0; i2 <= nPowerErrors; i2++ ) powerErrors[i2] = 0;

    for( i1 = 0; i1 < nData; i1++ ) {
        status = nf_exponentialIntegral( &smr, data[i1].n, data[i1].x, &En );
        r = 1;
        if( data[i1].En != 0. ) r = En / data[i1].En - 1;
        if( ( r > 3e-14 ) || ( status != nfu_Okay ) ) {
            printf( "%3d %.17e %.17e %.17e %+.3e  %d\n", data[i1].n, data[i1].x, data[i1].En, En, r, status );
            if( status != nfu_Okay ) nfut_printSMRError2p( &smr, "Via." );
            counts++;
        }
        for( i2 = 0, r2 = 1e-16; i2 < nPowerErrors; i2++, r2 *= 10. ) {
            if( fabs( r ) < r2 ) break;
        }
        powerErrors[i2]++;
    }
    if( info ) {
        printf( "relative" );
        for( i2 = 0; i2 <= nPowerErrors; i2++ ) printf( " %7d", -i2 - 6 );
        printf( "\n" );
        printf( "error:  " );
        for( i2 = 0; i2 <= nPowerErrors; i2++ ) printf( "%s%5d  %s", ( ( i2 < 4 ) ? " " : "" ), 10, ( ( i2 >= 4 ) ? " " : "" ) );
        printf( "\n" );

        printf( "--------" );
        for( i2 = 0; i2 <= nPowerErrors; i2++ ) printf( "--------" );
        printf( "\n" );

        printf( "counts: " );
        for( i2 = nPowerErrors; i2 >= 0; i2-- ) printf( " %7d", powerErrors[i2] );
        printf( "\n" );
    }
    exit( counts );
}
